index.js

```diff
@@ -16,159 +16,225 @@ const isDeflate = require('is-deflate')
+      // the reply in this case
+      buildRouteCompress(fastify, globalParams, routeOptions, true)
+    }
+  })
+
+  next()
+}
+
+function processParams (opts) {
+  if (!opts) {
+    return
   }
 
-  const onUnsupportedEncoding = opts.onUnsupportedEncoding
-  const inflateIfDeflated = opts.inflateIfDeflated === true
-  const threshold = typeof opts.threshold === 'number' ? opts.threshold : 1024
-  const compressibleTypes = opts.customTypes instanceof RegExp ? opts.customTypes : /^text\/|\+json$|\+text$|\+xml$|octet-stream$/
-  const compressStream = {
+  const params = {
+    global: (typeof opts.global === 'boolean') ? opts.global : true
+  }
+
+  params.onUnsupportedEncoding = opts.onUnsupportedEncoding
+  params.inflateIfDeflated = opts.inflateIfDeflated === true
+  params.threshold = typeof opts.threshold === 'number' ? opts.threshold : 1024
+  params.compressibleTypes = opts.customTypes instanceof RegExp ? opts.customTypes : /^text\/|\+json$|\+text$|\+xml$|octet-stream$/
+  params.compressStream = {
     gzip: (opts.zlib || zlib).createGzip || zlib.createGzip,
     deflate: (opts.zlib || zlib).createDeflate || zlib.createDeflate
   }
-  const uncompressStream = {
+  params.uncompressStream = {
     gzip: (opts.zlib || zlib).createGunzip || zlib.createGunzip,
     deflate: (opts.zlib || zlib).createInflate || zlib.createInflate
   }
 
   const supportedEncodings = ['gzip', 'deflate', 'identity']
   if (opts.brotli) {
-    compressStream.br = opts.brotli.compressStream
+    params.compressStream.br = opts.brotli.compressStream
     supportedEncodings.unshift('br')
   } else if (zlib.createBrotliCompress) {
-    compressStream.br = zlib.createBrotliCompress
+    params.compressStream.br = zlib.createBrotliCompress
     supportedEncodings.unshift('br')
   }
 
-  if (opts.encodings && opts.encodings.length < 1) {
-    next(new Error('The `encodings` option array must have at least 1 item.'))
-  }
-
-  const encodings = Array.isArray(opts.encodings)
+  params.encodings = Array.isArray(opts.encodings)
     ? supportedEncodings
       .filter(encoding => opts.encodings.includes(encoding))
       .sort((a, b) => opts.encodings.indexOf(a) - supportedEncodings.indexOf(b))
     : supportedEncodings
 
-  if (encodings.length < 1) {
-    next(new Error('None of the passed `encodings` were supported — compression not possible.'))
+  return params
+}
+
+function buildRouteCompress (fastify, params, routeOptions, decorateOnly) {
+  // In order to provide a compress method with the same parameter set as the route itself has
+  // we do the decorate the reply at the start of the request
+  if (Array.isArray(routeOptions.onRequest)) {
+    routeOptions.onRequest.push(onRequest)
+  } else if (typeof routeOptions.onRequest === 'function') {
+    routeOptions.onRequest = [routeOptions.onRequest, onRequest]
+  } else {
+    routeOptions.onRequest = [onRequest]
   }
 
-  next()
+  function onRequest (req, reply, next) {
+    reply.compress = compress(params)
```

Comment:
Given that `params` will not change, can you cache the output of `compress(params)` outside of the `onRequest` hook?

Reply:
👍